### PR TITLE
Nomad terraform config for enterprise binaries

### PR DESCRIPTION
**Pull request by Omar-Khawaja**
This PR is an attempt as solving the issue of using the terraform config with Nomad enterprise binaries for our team that are in restricted S3 buckets. A user spinning up a cluster in their AWS environment will be able to have the nodes read from S3 buckets that have given their AWS account the permission to do so.

This PR adds S3 GetObject permission to the IAM policy attached to all the nodes and adds some logic in the post-provisioning script when downloading/installing the nomad enterprise binary specified by the user. If this is a normal, open source binary from our website or a publicly accessible s3 bucket, nothing changes for the user. If, however, the user specifies an s3 link to a restricted bucket as a value to the **nomad_binary** variable (example: nomad_binary = "s3://my-bucket/my-enterprise-binary"), this configuration will install that.

Please note: you still have to set up an s3 bucket with an appropriate policy that will allow nodes in the AWS account to read from it, so these nodes will not be able to just read from any bucket in the same (or another) AWS account. See below for an example of how you would need to set up an s3 bucket to let these nodes read from it:

```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "Get object from S3 bucket",
            "Effect": "Allow",
            "Principal": {
                "AWS": "arn:aws:iam::<your-aws-account>:root"
            },
            "Action": "s3:*",
            "Resource": [
                "arn:aws:s3:::<your-s3-bucket>",
                "arn:aws:s3:::<your-s3-bucket/*>"
            ]
        }
    ]
}
```

### CODE REVIEW

File: terraform/aws/README.md [hunk 1]
```diff
@@ -57,7 +57,7 @@ variable like so:
 
 ```bash
 region                  = "us-east-1"
-ami                     = "ami-066a7f2ffac02f833"
+ami                     = "ami-05847887cacd41419"
```
Comment (notnoop) [inline terraform/aws/README.md]: What changed in the ami?  Was it a routine refresh as I don't quite see packer related changes here.

File: terraform/shared/scripts/client.sh [hunk 2]
```diff
@@ -30,14 +30,23 @@ sudo cp $CONFIGDIR/consul_$CLOUD.service /etc/systemd/system/consul.service
...
-  curl -L $NOMAD_BINARY > nomad.zip
-  sudo unzip -o nomad.zip -d /usr/local/bin
-  sudo chmod 0755 /usr/local/bin/nomad
-  sudo chown root:root /usr/local/bin/nomad
+        curl -L $NOMAD_BINARY > nomad.zip
+        sudo unzip -o nomad.zip -d /usr/local/bin
+        sudo chmod 0755 /usr/local/bin/nomad
+        sudo chown root:root /usr/local/bin/nomad
+elif [[ "$(which aws) s3 cp $NOMAD_BINARY nomad.zip --dryrun" ]]; then
+        $(which aws) s3 cp $NOMAD_BINARY nomad.zip
```
Comment (notnoop) [inline terraform/shared/scripts/client.sh]: I would suggest reworking this a bit: I would start with S3 and match over BINARY path, and keep the common code outside the if statement, maybe something like:

```bash
if [[ "$NOMAD_BINARY" = s3:* ]]; then
  s3 cp $NOMAD_BINARY nomad.zip
elif [[ `wget -S --spider $NOMAD_BINARY  2>&1 | grep 'HTTP/1.1 200 OK'` ]]; then
  curl -L $NOMAD_BINARY > nomad.zip
fi

sudo unzip -o nomad.zip -d /usr/local/bin
sudo chmod 0755 /usr/local/bin/nomad
sudo chown root:root /usr/local/bin/nomad
```

I find the use of wget and curl here odd but it's not quite related to this code.

File: terraform/aws/README.md (hunk 1 above)
Comment (Omar-Khawaja) [reply terraform/aws/README.md]: So the [user_data](https://github.com/hashicorp/nomad/tree/master/terraform/aws/env/us-east) scripts call the [setup](https://github.com/hashicorp/nomad/tree/master/terraform/shared/scripts) which are packed into the image. This is where I made sure the AWS CLI is installed and incorporated some [logic](https://github.com/hashicorp/nomad/blob/nomad-tf-config-ent/terraform/shared/scripts/client.sh#L45-L49) for pulling down the enterprise binaries form an S3 bucket.

File: terraform/shared/scripts/client.sh (hunk 2 above)
Comment (Omar-Khawaja) [reply terraform/shared/scripts/client.sh]: Ah yes excellent suggestion. Thank you!

**Comment by Omar-Khawaja** [comment]
Will re-visit this topic

**Pull request closed without merging**